library/core/src/num/uint_macros.rs

```diff
@@ -611,6 +611,155 @@ macro_rules! uint_impl {
+        }
+
+        /// Returns the base 2 logarithm of the number.
+        ///
+        /// # Panics
+        ///
+        /// Panics if the number is negative or zero.
+        ///
+        /// # Examples
+        ///
+        /// ```
+        /// #![feature(int_log)]
+        #[doc = concat!("assert_eq!(2", stringify!($SelfT), ".log2(), 1);")]
+        /// ```
+        #[unstable(feature = "int_log", issue = "70887")]
+        #[must_use = "this returns the result of the operation, \
+                        without modifying the original"]
+        #[inline]
+        pub fn log2(self) -> Self {
+            self.checked_log2().unwrap()
+        }
+
+        /// Returns the base 10 logarithm of the number.
+        ///
+        /// # Panics
+        ///
+        /// Panics if the number is negative or zero.
+        ///
+        /// # Example
+        ///
+        /// ```
+        /// #![feature(int_log)]
+        #[doc = concat!("assert_eq!(10", stringify!($SelfT), ".log10(), 1);")]
+        /// ```
+        #[unstable(feature = "int_log", issue = "70887")]
+        #[must_use = "this returns the result of the operation, \
+                        without modifying the original"]
+        #[inline]
+        pub fn log10(self) -> Self {
+            self.checked_log10().unwrap()
+        }
+
+        /// Returns the logarithm of the number with respect to an arbitrary base.
+        ///
+        /// Returns `None` if the number is zero, or if the base is not at least 2.
+        ///
+        /// This method may not be optimized owing to implementation details;
+        /// `self.checked_log2()` can produce results more efficiently for base 2, and
+        /// `self.checked_log10()` can produce results more efficiently for base 10.
+        ///
+        /// # Examples
+        ///
+        /// ```
+        /// #![feature(int_log)]
+        #[doc = concat!("assert_eq!(5", stringify!($SelfT), ".checked_log(5), Some(1));")]
+        /// ```
+        #[unstable(feature = "int_log", issue = "70887")]
+        #[must_use = "this returns the result of the operation, \
+                        without modifying the original"]
+        #[inline]
+        pub fn checked_log(self, base: Self) -> Option<Self> {
+            // SAFETY: We check the input to this is always positive.
+            let logb2 = |x: Self| unsafe { intrinsics::ctlz_nonzero(1 as Self) - intrinsics::ctlz_nonzero(x) };
+
+            if self <= 0 || base <= 1 {
+                None
+            } else {
+                let mut n = 0;
+                let mut r = self;
+
+                // Optimization for 128 bit wide integers.
+                if Self::BITS == 128 {
+                    let b = logb2(self) / (logb2(base) + 1);
+                    n += b;
+                    r /= base.pow(b as u32);
+                }
+
+                while r >= base {
+                    r /= base;
+                    n += 1;
+                }
```

Comment:
I guess (haven't checked) that this optimizes to multiplications for a fixed base anyway, but wouldn't it be faster in the general case to use a loop of multiplications (an unrolled power calculation basically) instead?

Reply:
I'm not super strong on math and algorithms, so I'd like to leave the implementation as-is for this PR, and defer further optimizations to follow-up PRs. This has undergone several rounds of optimizations already, so I think it's already faster than most ad-hoc impls would be.

Reply:
Sure, sounds fine.

Reply:
I understand that you're leaving the implementation as is while the code goes through the process and I appreciate that. When you're ready for an optimized version, please see the function checked_log_with_largest_squaring here for a possible implementation: https://gist.github.com/rust-play/cee6917b15c465635d31af30f9f60599

You can read more about the math behind it at a blog post I wrote:
https://secondgenerationprogrammer.wordpress.com/2016/08/10/egyptian-logarithms/